線幅のサポートを追加 / ベースとするシェーダーを変更できるように

hiyakake opened the pull request:
## 概要
このプルリクエストでは、QvPenエクスポーターとインポーターに線幅のサポートを追加しました。
これにより、エクスポートされたQvPenのストロークを元の線幅で正確に復元できるようになります。

また、インポート時に利用するシェーダーを変更できるようにしています。

## 変更内容

### エクスポーター側の変更
- エクスポート時にQvPenの線幅情報をJSONに含めるよう対応
- 線幅が0の場合のフォールバックメカニズムを追加

### インポーター側の変更
- JSONから線幅情報を読み取る機能を追加
- 線幅に応じたマテリアルバリアントを自動生成するシステムを実装
- プレビュー画面に線幅情報の表示を追加

## 技術的な詳細
QvPenでは、線の太さはLineRenderer自体の`widthMultiplier`ではなく、シェーダーの`_Width`プロパティで制御されています。そのため、インポート時には線幅ごとに異なるマテリアルバリアントを生成し、それをLineRendererに適用する方式を採用しました。これにより元のQvPenでの線幅を忠実に再現できると思います。

## テスト結果
様々な太さのペンで描画したストロークをエクスポート・インポートし、元の描画と視覚的に一致することを確認しました。

hiyakake committed:
ペンの太さ情報をエクスポート処理に追加

- LineRendererの太さ情報をJSONエクスポートに
- デフォルトのペンの太さを設定するためのシリアライズフィールドを追加
- ペンマネージャーから太さ情報を取得する処理を実装
- 太さが0の場合にデフォルト値を使用するロジックを追加

hiyakake committed:
QvPenインポーターに線の太さ対応機能を追加

- 線の太さに応じたマテリアルの動的生成と管理を実装
- ベースマテリアルの自動検索と手動選択機能を追加
- インポート時に線の太さ情報を反映するよう修正
- マテリアル保存用フォルダを自動作成
- インポート時のUI表示を改善し、詳細な情報を提供

Dolphiiiin commented on Packages/com.dolphiiiin.qvpen-exporter/Editor/QvPenImporter.cs:
```diff
@@ -231,11 +361,28 @@ private void CreateLineRenderers()
 
                         lineRenderer.positionCount = pointCount;
                         lineRenderer.SetPositions(positions);
+                        
```
QvPenImporterで同じcolorのペンで異なる線の太さがjson内で存在する場合に、線の太さのプレビュー表示が正しく表記されません。

問題が発生するjsonファイルです。このjsonファイル内では以下の内容でペンがエクスポートされています。
- color: 000000, width: 0.02
- color: FF0000, width: 0.005, 0.2
- color: FF8000, width: 0.005
[export_2025_03_12_20_10_49.json](https://github.com/user-attachments/files/19209811/export_2025_03_12_20_10_49.json)
QvPenImporterでのGUI上ではデータのうち一種の太さしかプレビューが行われません。
![image](https://github.com/user-attachments/assets/9648339b-9824-4aa9-9450-f16fbec56f46)

QvPenExporterでのInkPoolのセットアップはユーザーによって行われることを考慮して、複数ペンの太さが混在する可能性が考えられます。
(現在ToDoでInkPoolを自動セットアップするための、Editorスクリプトを追加することが計画されています。)

確認していただけますと幸いです。

hiyakake replied:
レビューありがとうございます。
以下のように修正しましたので再度チェックをお願いします。

## 概要
同じ色でも異なる線の太さを持つペンデータが混在する場合の処理を改善しました。
同時にグルーピングの方法を改善しています。

VRChat上で描いたもの
![image](https://github.com/user-attachments/assets/5153696c-c4a8-406a-b544-0a5d0cd962a1)
イラスト：クミンさん

改善前
![image](https://github.com/user-attachments/assets/be28b20a-c5cc-481d-9cff-2c56b92e67fa)

改善後
![image](https://github.com/user-attachments/assets/3ce74f1d-b9f8-4fc1-8c58-0e742162035e)

## 変更内容

### 1. オブジェクトのグループ化方法の改善
- 色のみでグループ化していた仕様を、色と線の太さの組み合わせでグループ化するように変更
- グループキーのフォーマットを `{colorHex}_width{width:F4}` の形式に統一
- グループオブジェクト名を `Group_{colorHex}_{widthStr}` の形式に変更し、より分かりやすく

### 2. プレビュー表示の強化
- 「色ごとのグループ」と「色と太さの組み合わせごとのグループ」の2セクションを追加
- 各グループのライン数を明確に表示
- 線の太さの視覚的表現を改善
- 合計ライン数の表示を追加

### 3. データ構造の最適化
- グループ化のためのデータ構造を改善
- 色と太さの組み合わせによる効率的なグループ管理を実装
- デフォルト値（width = 0.005f）の適切な処理

## 技術的な詳細
- `CreateLineRenderers`メソッドを色と太さの組み合わせでグループ化するように再構築
- プレビュー表示用のデータ構造を2段階（色のみ/色と太さ）に分離
- LineRendererオブジェクトの命名規則を統一化

Dolphiiiin commented on Packages/com.dolphiiiin.qvpen-exporter/Runtime/UdonScript/QvPenExporter.cs:
確認しました。こちらは修正なしで問題ありません 👍

Dolphiiiin commented:
プルリクエストありがとうございます！レビューの方をさせていただきましたので、ご確認ください 🙏

hiyakake commented:
レビューありがとうございます。
問題は起きなかったということですか？

hiyakake committed:
QvPenインポーターに色と太さの組み合わせでのグループ化機能を追加

- 色ごとにグループ化したデータ構造を実装
- 色と太さの組み合わせでのグループ化を追加し、UI表示を改善
- 各グループのライン数や太さを詳細に表示する機能を追加
- デフォルトの太さを使用するロジックを強化し、エクスポート処理を最適化

Dolphiiiin reviewed (approved):
修正ありがとうございます 🙏確認しましたので、こちらでマージさせていただきます！
